File: runtime/CSharp/Antlr4.Runtime/Tree/TerminalNodeImpl.cs
// Copyright (c) Terence Parr, Sam Harwell. All Rights Reserved.
// Licensed under the BSD License. See LICENSE.txt in the project root for license information.

/*
* Copyright (c) 2012 The ANTLR Project. All rights reserved.
* Use of this file is governed by the BSD-3-Clause license that
* can be found in the LICENSE.txt file in the project root.
*/
using Antlr4.Runtime;
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Tree
{
    public class TerminalNodeImpl : ITerminalNode
    {
        public IToken symbol;

        public IRuleNode parent;

        public TerminalNodeImpl(IToken symbol)
        {
            this.symbol = symbol;
        }

        public virtual IParseTree GetChild(int i)
        {
            return null;
        }

        public virtual IToken Symbol
        {
            get
            {
                return symbol;
            }
        }

        public virtual IRuleNode Parent
        {
            get
            {
                return parent;
            }
        }

        public virtual IToken Payload
        {
            get
            {
                return symbol;
            }
        }

        public virtual Interval SourceInterval
        {
            get
            {
                if (symbol != null)
                {
                    int tokenIndex = symbol.TokenIndex;
                    return new Interval(tokenIndex, tokenIndex);
                }
                return Interval.Invalid;
            }
        }

        public virtual int ChildCount
        {
            get
            {
                return 0;
            }
        }

        public virtual T Accept<T, _T1>(IParseTreeVisitor<_T1> visitor)
            where _T1 : T
        {
            return visitor.VisitTerminal(this);
        }

        public virtual string GetText()
        {
            if (symbol != null)
            {
                return symbol.Text;
            }
            return null;
        }

        public virtual string ToStringTree(Parser parser)
        {
            return ToString();
        }

        public override string ToString()
        {
            if (symbol != null)
            {
                if (symbol.Type == TokenConstants.Eof)
                {
                    return "<EOF>";
                }
                return symbol.Text;
            }
            else
            {
                return "<null>";
            }
        }

        public virtual string ToStringTree()
        {
            return ToString();
        }
    }
}

File: reference/antlr4/runtime/Java/src/org/antlr/v4/runtime/tree/TerminalNodeImpl.java
/*
 * Copyright (c) 2012 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD-3-Clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.tree;

import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.misc.Interval;

import java.util.List;

public class TerminalNodeImpl implements TerminalNode {
	public Token symbol;
	public RuleNode parent;

	public TerminalNodeImpl(Token symbol) {	this.symbol = symbol;	}

	@Override
	public ParseTree getChild(int i) {return null;}

	@Override
	public Token getSymbol() {return symbol;}

	@Override
	public RuleNode getParent() { return parent; }

	@Override
	public Token getPayload() { return symbol; }

	@Override
	public Interval getSourceInterval() {
		if (symbol != null) {
			int tokenIndex = symbol.getTokenIndex();
			return new Interval(tokenIndex, tokenIndex);
		}

		return Interval.INVALID;
	}

	@Override
	public int getChildCount() { return 0; }

	@Override
	public <T> T accept(ParseTreeVisitor<? extends T> visitor) {
		return visitor.visitTerminal(this);
	}

	@Override
	public String getText() {
		if (symbol != null) {
			return symbol.getText();
		}

		return null;
	}

	@Override
	public String toStringTree(Parser parser) {
		return toString();
	}

	@Override
	public String toString() {
		if (symbol != null) {
			if ( symbol.getType() == Token.EOF ) {
				return "<EOF>";
			}

			return symbol.getText();
		}
		else {
			return "<null>";
		}
	}

	@Override
	public String toStringTree() {
		return toString();
	}
}
